source/ring_buffer.c

```diff
@@ -0,0 +1,206 @@
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/ring_buffer.h>
+
+int aws_ring_buffer_init(struct aws_ring_buffer *ring_buf, struct aws_allocator *allocator, size_t size) {
+    AWS_ZERO_STRUCT(*ring_buf);
+
+    ring_buf->allocation = aws_mem_acquire(allocator, size);
+
+    if (!ring_buf->allocation) {
+        return AWS_OP_ERR;
+    }
+
+    ring_buf->allocator = allocator;
+    aws_atomic_store_ptr(&ring_buf->head, ring_buf->allocation);
+    aws_atomic_store_ptr(&ring_buf->tail, ring_buf->allocation);
+    ring_buf->allocation_end = ring_buf->allocation + size + 1;
+
+    return AWS_OP_SUCCESS;
+}
+
+void aws_ring_buffer_clean_up(struct aws_ring_buffer *ring_buf) {
+    if (ring_buf->allocation) {
+        aws_mem_release(ring_buf->allocator, ring_buf->allocation);
+    }
+
+    AWS_ZERO_STRUCT(*ring_buf);
+}
+
+int aws_ring_buffer_acquire(struct aws_ring_buffer *ring_buf, size_t requested_size, struct aws_byte_buf *dest) {
+    if (requested_size == 0) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    uint8_t *tail_cpy = aws_atomic_load_ptr(&ring_buf->tail);
+    uint8_t *head_cpy = aws_atomic_load_ptr(&ring_buf->head);
+
+    /* this branch is, we don't have any vended buffers. */
+    if (head_cpy == tail_cpy) {
+        size_t ring_space = ring_buf->allocation_end - ring_buf->allocation;
+        size_t current_space = ring_buf->allocation_end - head_cpy;
+
+        if (requested_size > ring_space) {
+            return aws_raise_error(AWS_ERROR_NO_AVAILABLE_BUFFERS);
+        }
+
+        if (current_space >= requested_size) {
+            aws_atomic_store_ptr(&ring_buf->head, head_cpy + requested_size);
+            *dest = aws_byte_buf_from_empty_array(head_cpy, requested_size);
+            return AWS_OP_SUCCESS;
+        }
+
+        aws_atomic_store_ptr(&ring_buf->head, ring_buf->allocation + requested_size);
+        aws_atomic_store_ptr(&ring_buf->tail, ring_buf->allocation);
+        *dest = aws_byte_buf_from_empty_array(ring_buf->allocation, requested_size);
+        return AWS_OP_SUCCESS;
+    }
+
+    /* you'll constantly bounce between the next two branches as the ring buffer is traversed. */
+    /* after N + 1 wraps */
+    if (tail_cpy > head_cpy) {
+        size_t space = tail_cpy - head_cpy - 1;
+
+        if (space >= requested_size) {
+            aws_atomic_store_ptr(&ring_buf->head, head_cpy + requested_size);
+            *dest = aws_byte_buf_from_empty_array(head_cpy, requested_size);
+            return AWS_OP_SUCCESS;
+        }
+        /* After N wraps */
+    } else if (tail_cpy < head_cpy) {
+        /* prefer the head space for efficiency. */
+        if ((size_t)(ring_buf->allocation_end - head_cpy) > requested_size) {
+            aws_atomic_store_ptr(&ring_buf->head, head_cpy + requested_size);
+            *dest = aws_byte_buf_from_empty_array(head_cpy, requested_size);
+            return AWS_OP_SUCCESS;
```

Comment:
[trivial] I **think** this could be: `if ((allocation_end - head_cpy) >= requested_size)`
can't we vend all the way to the end of the buffer, without leaving a gap, since we'd still end up with head != tail?

Reply:
allocation_end is one past the end of the allocation

Reply:
say the ring has 2 bytes, and we've vended 1, so:
allocation: 0
allocation_end: 2
tail: 0
head: 1
request: 1

We *could* vend that last byte, but with the code as written: `if ((2 - 1) > 1)` we won't vend that last byte

Reply:
updated